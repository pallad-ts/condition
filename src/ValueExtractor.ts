import * as is from 'predicates';
import {DENORMALIZE, NORMALIZE} from "./global";

export abstract class ValueExtractor<T> {
    abstract extract(value: T): Promise<any> | any;

    abstract toString(): string;
}

export namespace ValueExtractor {

    export class Property<T> extends ValueExtractor<T> {
        readonly path: Array<string | symbol>;

        constructor(...path: Property.Path) {
            if (path.length <= 0) {
                throw new Error('Property path must be longer than 0');
            }

            super();
            this.path = path;
            Object.freeze(this);
        }

        static create(...path: Property.Path) {
            return new Property(...path);
        }

        async extract(value: T): Promise<any> {
            let currentValue: any = value;
            for (const part of this.path) {
                if (is.object(currentValue)) {
                    currentValue = await currentValue[part];
                } else {
                    return undefined;
                }
            }
            return currentValue;
        }

        toString() {
            return `property "${this.path.join('.')}"`;
        }

        [NORMALIZE]() {
            return this.path;
        }

        static [DENORMALIZE](data: string[]) {
            return new Property(...data);
        }
    }

    export namespace Property {
        export type Path = Part[];
        export type Part = symbol | string;
    }
}
