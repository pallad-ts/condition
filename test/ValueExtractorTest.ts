import {ValueExtractor} from "@src/ValueExtractor";

describe('ValueExtractor', () => {
    describe('Property', () => {
        const VALUE_ON_PATH = 'test';

        it.each<[ValueExtractor<any>, any, string]>([
            [
                new ValueExtractor.Property('foo'),
                {foo: 'test'},
                'property "foo"'
            ],
            [
                new ValueExtractor.Property('foo', 'bar'),
                {
                    foo: {
                        bar :VALUE_ON_PATH
                    }
                },
                'property "foo.bar"'
            ],
            [
                new ValueExtractor.Property('foo', 'bar'),
                {
                    get foo() {
                        return Promise.resolve({
                            bar: VALUE_ON_PATH
                        })
                    }
                },
                'property "foo.bar"'
            ],
            [
                new ValueExtractor.Property('foo', 'bar', 'test'),
                {
                    get foo() {
                        return Promise.resolve({
                            get bar() {
                                return Promise.resolve({
                                    test: VALUE_ON_PATH
                                })
                            }
                        })
                    }
                },
                'property "foo.bar.test"'
            ]
        ])('retrieves value at given key path with promise supporte', async (extractor, validValue, desc) => {
            await expect(extractor.extract(validValue))
                .resolves
                .toEqual(VALUE_ON_PATH);

            await expect(extractor.extract({}))
                .resolves
                .toBeUndefined();

            expect(extractor.toString())
                .toEqual(desc);
        });
    });

    it('throw an error when too short path provided', () => {
        expect(() => {
            new ValueExtractor.Property()
        })
            .toThrowErrorMatchingSnapshot();
    });

    it('creating through factory', () => {
        const path = ['foo', 'bar'];
        expect(new ValueExtractor.Property(...path))
            .toEqual(ValueExtractor.Property.create(...path));
    });
});