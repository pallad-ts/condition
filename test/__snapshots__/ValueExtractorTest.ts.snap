// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValueExtractor throw an error when too short path provided 1`] = `"Property path must be longer than 0"`;
